File: scripts/r/visualizations.R
# Business Analytics - Data Visualization Script (R)
# Purpose: Create business visualizations using ggplot2
# Author: Business Analytics Team
# Last Updated: 2024

# Load required libraries
library(ggplot2)
library(dplyr)
library(tidyr)
library(scales)
library(readr)

# Custom theme for business charts
theme_business <- function() {
  theme_minimal() +
    theme(
      plot.title = element_text(size = 14, face = "bold", hjust = 0.5),
      plot.subtitle = element_text(size = 10, hjust = 0.5, color = "gray40"),
      axis.title = element_text(size = 10, face = "bold"),
      axis.text = element_text(size = 9),
      legend.position = "bottom",
      legend.title = element_text(face = "bold"),
      panel.grid.minor = element_blank(),
      plot.margin = margin(10, 10, 10, 10)
    )
}

# Load cleaned data
data_path <- "data/processed/sales_data_cleaned.csv"

if (file.exists(data_path)) {
  sales_data <- read_csv(data_path, show_col_types = FALSE)
} else {
  cat("Error: Cleaned data not found. Run data_cleaning.R first.\n")
  stop()
}

# 1. Sales by Region (Bar Chart)
plot_sales_by_region <- function(data) {
  region_summary <- data %>%
    group_by(region) %>%
    summarise(
      total_sales = sum(sales_amount),
      avg_transaction = mean(sales_amount),
      .groups = 'drop'
    ) %>%
    arrange(desc(total_sales))
  
  p <- ggplot(region_summary, aes(x = reorder(region, total_sales), y = total_sales)) +
    geom_bar(stat = "identity", fill = "#2E86AB", alpha = 0.8) +
    geom_text(aes(label = dollar(total_sales)), 
              hjust = -0.1, size = 3.5, fontface = "bold") +
    coord_flip() +
    scale_y_continuous(labels = dollar_format(), expand = expansion(mult = c(0, 0.15))) +
    labs(
      title = "Total Sales by Region",
      subtitle = "Q1 2024 Performance",
      x = "Region",
      y = "Total Sales ($)"
    ) +
    theme_business()
  
  return(p)
}

# 2. Sales Trend Over Time (Line Chart)
plot_sales_trend <- function(data) {
  daily_sales <- data %>%
    group_by(date) %>%
    summarise(daily_total = sum(sales_amount), .groups = 'drop')
  
  p <- ggplot(daily_sales, aes(x = date, y = daily_total)) +
    geom_line(color = "#A23B72", size = 1) +
    geom_point(color = "#A23B72", size = 2) +
    geom_smooth(method = "loess", se = TRUE, color = "#F18F01", alpha = 0.2) +
    scale_y_continuous(labels = dollar_format()) +
    scale_x_date(date_breaks = "1 week", date_labels = "%b %d") +
    labs(
      title = "Daily Sales Trend",
      subtitle = "With trend line (LOESS smoothing)",
      x = "Date",
      y = "Daily Sales ($)"
    ) +
    theme_business() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  
  return(p)
}

# 3. Sales by Product Category (Pie Chart alternative - Donut)
plot_sales_by_category <- function(data) {
  category_summary <- data %>%
    group_by(product_category) %>%
    summarise(total_sales = sum(sales_amount), .groups = 'drop') %>%
    mutate(
      percentage = total_sales / sum(total_sales) * 100,
      label = paste0(product_category, "\n", dollar(total_sales), 
                    "\n(", round(percentage, 1), "%)")
    )
  
  p <- ggplot(category_summary, aes(x = "", y = total_sales, fill = product_category)) +
    geom_bar(stat = "identity", width = 1, color = "white", size = 2) +
    coord_polar("y", start = 0) +
    scale_fill_manual(values = c("#2E86AB", "#A23B72", "#F18F01")) +
    labs(
      title = "Sales Distribution by Product Category",
      subtitle = "Q1 2024",
      fill = "Category"
    ) +
    theme_void() +
    theme(
      plot.title = element_text(size = 14, face = "bold", hjust = 0.5),
      plot.subtitle = element_text(size = 10, hjust = 0.5, color = "gray40"),
      legend.position = "right"
    )
  
  return(p)
}

# 4. Monthly Sales Comparison (Grouped Bar Chart)
plot_monthly_comparison <- function(data) {
  monthly_category <- data %>%
    group_by(month, product_category) %>%
    summarise(total_sales = sum(sales_amount), .groups = 'drop')
  
  p <- ggplot(monthly_category, aes(x = month, y = total_sales, fill = product_category)) +
    geom_bar(stat = "identity", position = "dodge", alpha = 0.8) +
    scale_y_continuous(labels = dollar_format()) +
    scale_fill_manual(values = c("#2E86AB", "#A23B72", "#F18F01")) +
    labs(
      title = "Monthly Sales by Product Category",
      subtitle = "Q1 2024 Breakdown",
      x = "Month",
      y = "Total Sales ($)",
      fill = "Category"
    ) +
    theme_business() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  
  return(p)
}

# 5. Sales Rep Performance (Horizontal Bar Chart)
plot_rep_performance <- function(data) {
  rep_summary <- data %>%
    group_by(sales_rep) %>%
    summarise(
      total_sales = sum(sales_amount),
      num_transactions = n(),
      avg_sale = mean(sales_amount),
      .groups = 'drop'
    ) %>%
    arrange(desc(total_sales))
  
  p <- ggplot(rep_summary, aes(x = reorder(sales_rep, total_sales), y = total_sales)) +
    geom_bar(stat = "identity", fill = "#F18F01", alpha = 0.8) +
    geom_text(aes(label = paste0(dollar(total_sales), "\n(", num_transactions, " sales)")), 
              hjust = -0.1, size = 3) +
    coord_flip() +
    scale_y_continuous(labels = dollar_format(), expand = expansion(mult = c(0, 0.15))) +
    labs(
      title = "Sales Representative Performance",
      subtitle = "Total Sales and Number of Transactions",
      x = "Sales Rep",
      y = "Total Sales ($)"
    ) +
    theme_business()
  
  return(p)
}

# Generate and save all plots
save_all_plots <- function() {
  dir.create("output/figures", recursive = TRUE, showWarnings = FALSE)
  
  cat("Generating visualizations...\n")
  
  # Save plots
  ggsave("output/figures/sales_by_region.png", 
         plot_sales_by_region(sales_data), 
         width = 8, height = 6, dpi = 300)
  cat("✓ Saved: sales_by_region.png\n")
  
  ggsave("output/figures/sales_trend.png", 
         plot_sales_trend(sales_data), 
         width = 10, height = 6, dpi = 300)
  cat("✓ Saved: sales_trend.png\n")
  
  ggsave("output/figures/sales_by_category.png", 
         plot_sales_by_category(sales_data), 
         width = 8, height = 6, dpi = 300)
  cat("✓ Saved: sales_by_category.png\n")
  
  ggsave("output/figures/monthly_comparison.png", 
         plot_monthly_comparison(sales_data), 
         width = 10, height = 6, dpi = 300)
  cat("✓ Saved: monthly_comparison.png\n")
  
  ggsave("output/figures/rep_performance.png", 
         plot_rep_performance(sales_data), 
         width = 8, height = 6, dpi = 300)
  cat("✓ Saved: rep_performance.png\n")
  
  cat("\nAll visualizations saved to output/figures/\n")
}

# Main execution
if (!interactive()) {
  save_all_plots()
} else {
  cat("Available functions:\n")
  cat("  - plot_sales_by_region(sales_data)\n")
  cat("  - plot_sales_trend(sales_data)\n")
  cat("  - plot_sales_by_category(sales_data)\n")
  cat("  - plot_monthly_comparison(sales_data)\n")
  cat("  - plot_rep_performance(sales_data)\n")
  cat("  - save_all_plots() - Generate all plots\n")
}

File: scripts/python/visualizations.py
"""
Business Analytics - Data Visualization Script (Python)
Purpose: Create business visualizations using matplotlib and seaborn
Author: Business Analytics Team
Last Updated: 2024
"""

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from matplotlib.patches import Rectangle
import os

# Set style for professional-looking charts
sns.set_style("whitegrid")
plt.rcParams['figure.figsize'] = (10, 6)
plt.rcParams['font.size'] = 10
plt.rcParams['axes.titlesize'] = 14
plt.rcParams['axes.titleweight'] = 'bold'

# Configuration
PROCESSED_DATA_PATH = "data/processed/sales_data_cleaned.csv"
OUTPUT_PATH = "output/figures"

# Color palette
COLORS = {
    'primary': '#2E86AB',
    'secondary': '#A23B72',
    'accent': '#F18F01',
    'success': '#06A77D',
    'warning': '#F77F00'
}

def load_data():
    """Load cleaned sales data"""
    if not os.path.exists(PROCESSED_DATA_PATH):
        print(f"Error: Cleaned data not found at {PROCESSED_DATA_PATH}")
        print("Please run R data_cleaning.R first.")
        return None
    
    df = pd.read_csv(PROCESSED_DATA_PATH)
    df['date'] = pd.to_datetime(df['date'])
    print(f"✓ Loaded {len(df)} records for visualization")
    return df

def plot_revenue_heatmap(df):
    """Create heatmap of revenue by region and category"""
    fig, ax = plt.subplots(figsize=(10, 6))
    
    # Pivot data
    heatmap_data = df.pivot_table(
        values='sales_amount',
        index='region',
        columns='product_category',
        aggfunc='sum'
    )
    
    # Create heatmap
    sns.heatmap(heatmap_data, annot=True, fmt=',.0f', cmap='YlOrRd',
                cbar_kws={'label': 'Total Sales ($)'}, ax=ax)
    
    ax.set_title('Revenue Heatmap: Region × Product Category', pad=20)
    ax.set_xlabel('Product Category', fontweight='bold')
    ax.set_ylabel('Region', fontweight='bold')
    
    plt.tight_layout()
    return fig

def plot_sales_distribution(df):
    """Create distribution plot of sales amounts"""
    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(14, 6))
    
    # Histogram with KDE
    ax1.hist(df['sales_amount'], bins=30, color=COLORS['primary'], 
             alpha=0.7, edgecolor='black')
    ax1.axvline(df['sales_amount'].mean(), color=COLORS['warning'], 
                linestyle='--', linewidth=2, label=f"Mean: ${df['sales_amount'].mean():.2f}")
    ax1.axvline(df['sales_amount'].median(), color=COLORS['success'], 
                linestyle='--', linewidth=2, label=f"Median: ${df['sales_amount'].median():.2f}")
    ax1.set_xlabel('Sales Amount ($)', fontweight='bold')
    ax1.set_ylabel('Frequency', fontweight='bold')
    ax1.set_title('Sales Amount Distribution', pad=15)
    ax1.legend()
    ax1.grid(True, alpha=0.3)
    
    # Box plot by category
    df.boxplot(column='sales_amount', by='product_category', ax=ax2, 
               patch_artist=True, grid=True)
    ax2.set_xlabel('Product Category', fontweight='bold')
    ax2.set_ylabel('Sales Amount ($)', fontweight='bold')
    ax2.set_title('Sales Distribution by Category', pad=15)
    plt.suptitle('')  # Remove default title
    
    plt.tight_layout()
    return fig

def plot_time_series_advanced(df):
    """Create advanced time series visualization"""
    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(12, 10))
    
    # Daily sales with moving average
    daily_sales = df.groupby('date')['sales_amount'].sum().reset_index()
    daily_sales['MA_7'] = daily_sales['sales_amount'].rolling(window=7, min_periods=1).mean()
    
    ax1.plot(daily_sales['date'], daily_sales['sales_amount'], 
             color=COLORS['primary'], alpha=0.5, linewidth=1, label='Daily Sales')
    ax1.plot(daily_sales['date'], daily_sales['MA_7'], 
             color=COLORS['secondary'], linewidth=2.5, label='7-Day Moving Average')
    ax1.fill_between(daily_sales['date'], daily_sales['sales_amount'], 
                      alpha=0.2, color=COLORS['primary'])
    ax1.set_xlabel('Date', fontweight='bold')
    ax1.set_ylabel('Sales Amount ($)', fontweight='bold')
    ax1.set_title('Daily Sales Trend with Moving Average', pad=15)
    ax1.legend()
    ax1.grid(True, alpha=0.3)
    ax1.tick_params(axis='x', rotation=45)
    
    # Cumulative sales by category
    for category in df['product_category'].unique():
        cat_data = df[df['product_category'] == category].sort_values('date')
        cumulative = cat_data.groupby('date')['sales_amount'].sum().cumsum()
        ax2.plot(cumulative.index, cumulative.values, linewidth=2.5, 
                label=category, marker='o', markersize=3)
    
    ax2.set_xlabel('Date', fontweight='bold')
    ax2.set_ylabel('Cumulative Sales ($)', fontweight='bold')
    ax2.set_title('Cumulative Sales by Product Category', pad=15)
    ax2.legend()
    ax2.grid(True, alpha=0.3)
    ax2.tick_params(axis='x', rotation=45)
    
    plt.tight_layout()
    return fig

def plot_regional_comparison(df):
    """Create comprehensive regional comparison"""
    fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2, figsize=(14, 10))
    
    # 1. Total sales by region (bar chart)
    region_sales = df.groupby('region')['sales_amount'].sum().sort_values(ascending=True)
    colors_bar = [COLORS['primary'] if x == region_sales.max() else COLORS['secondary'] 
                  for x in region_sales]
    region_sales.plot(kind='barh', ax=ax1, color=colors_bar, alpha=0.8)
    ax1.set_xlabel('Total Sales ($)', fontweight='bold')
    ax1.set_ylabel('Region', fontweight='bold')
    ax1.set_title('Total Sales by Region', pad=15)
    ax1.grid(True, alpha=0.3, axis='x')
    
    # Add value labels
    for i, v in enumerate(region_sales.values):
        ax1.text(v + 1000, i, f'${v:,.0f}', va='center', fontweight='bold')
    
    # 2. Average transaction value by region
    region_avg = df.groupby('region')['sales_amount'].mean().sort_values(ascending=True)
    region_avg.plot(kind='barh', ax=ax2, color=COLORS['accent'], alpha=0.8)
    ax2.set_xlabel('Average Transaction ($)', fontweight='bold')
    ax2.set_ylabel('Region', fontweight='bold')
    ax2.set_title('Average Transaction Value by Region', pad=15)
    ax2.grid(True, alpha=0.3, axis='x')
    
    # 3. Number of transactions by region (pie chart)
    region_count = df.groupby('region').size()
    colors_pie = [COLORS['primary'], COLORS['secondary'], COLORS['accent'], COLORS['success']]
    ax3.pie(region_count.values, labels=region_count.index, autopct='%1.1f%%',
            startangle=90, colors=colors_pie, textprops={'fontweight': 'bold'})
    ax3.set_title('Transaction Distribution by Region', pad=15)
    
    # 4. Sales by region and month (stacked bar)
    region_month = df.pivot_table(values='sales_amount', index='month_num', 
                                    columns='region', aggfunc='sum')
    region_month.plot(kind='bar', stacked=True, ax=ax4, 
                      color=[COLORS['primary'], COLORS['secondary'], 
                            COLORS['accent'], COLORS['success']], alpha=0.8)
    ax4.set_xlabel('Month', fontweight='bold')
    ax4.set_ylabel('Total Sales ($)', fontweight='bold')
    ax4.set_title('Monthly Sales Trend by Region', pad=15)
    ax4.legend(title='Region', bbox_to_anchor=(1.05, 1), loc='upper left')
    ax4.grid(True, alpha=0.3, axis='y')
    ax4.tick_params(axis='x', rotation=0)
    
    plt.tight_layout()
    return fig

def plot_category_performance(df):
    """Create category performance dashboard"""
    fig = plt.figure(figsize=(14, 10))
    gs = fig.add_gridspec(3, 2, hspace=0.3, wspace=0.3)
    
    # 1. Revenue by category
    ax1 = fig.add_subplot(gs[0, :])
    category_revenue = df.groupby('product_category').agg({
        'sales_amount': 'sum',
        'units_sold': 'sum'
    }).sort_values('sales_amount', ascending=False)
    
    x = np.arange(len(category_revenue.index))
    width = 0.35
    
    ax1_twin = ax1.twinx()
    bars1 = ax1.bar(x - width/2, category_revenue['sales_amount'], width, 
                     label='Revenue', color=COLORS['primary'], alpha=0.8)
    bars2 = ax1_twin.bar(x + width/2, category_revenue['units_sold'], width, 
                          label='Units Sold', color=COLORS['accent'], alpha=0.8)
    
    ax1.set_xlabel('Product Category', fontweight='bold')
    ax1.set_ylabel('Revenue ($)', fontweight='bold', color=COLORS['primary'])
    ax1_twin.set_ylabel('Units Sold', fontweight='bold', color=COLORS['accent'])
    ax1.set_title('Category Performance: Revenue vs Units Sold', pad=15)
    ax1.set_xticks(x)
    ax1.set_xticklabels(category_revenue.index, rotation=0)
    ax1.tick_params(axis='y', labelcolor=COLORS['primary'])
    ax1_twin.tick_params(axis='y', labelcolor=COLORS['accent'])
    ax1.grid(True, alpha=0.3, axis='y')
    
    # Combine legends
    lines1, labels1 = ax1.get_legend_handles_labels()
    lines2, labels2 = ax1_twin.get_legend_handles_labels()
    ax1.legend(lines1 + lines2, labels1 + labels2, loc='upper left')
    
    # 2. Unit price analysis
    ax2 = fig.add_subplot(gs[1, 0])
    category_price = df.groupby('product_category')['unit_price'].agg(['mean', 'min', 'max'])
    category_price.plot(kind='bar', ax=ax2, color=[COLORS['primary'], COLORS['secondary'], COLORS['accent']], 
                        alpha=0.8)
    ax2.set_xlabel('Product Category', fontweight='bold')
    ax2.set_ylabel('Unit Price ($)', fontweight='bold')
    ax2.set_title('Unit Price Analysis by Category', pad=15)
    ax2.legend(['Average', 'Minimum', 'Maximum'])
    ax2.grid(True, alpha=0.3, axis='y')
    ax2.tick_params(axis='x', rotation=45)
    
    # 3. Category trend over time
    ax3 = fig.add_subplot(gs[1, 1])
    for category in df['product_category'].unique():
        cat_daily = df[df['product_category'] == category].groupby('date')['sales_amount'].sum()
        ax3.plot(cat_daily.index, cat_daily.values, linewidth=2, label=category, marker='o', markersize=2)
    ax3.set_xlabel('Date', fontweight='bold')
    ax3.set_ylabel('Daily Sales ($)', fontweight='bold')
    ax3.set_title('Daily Sales Trend by Category', pad=15)
    ax3.legend()
    ax3.grid(True, alpha=0.3)
    ax3.tick_params(axis='x', rotation=45)
    
    # 4. Category mix by region
    ax4 = fig.add_subplot(gs[2, :])
    category_region = df.pivot_table(values='sales_amount', index='region', 
                                      columns='product_category', aggfunc='sum')
    category_region_pct = category_region.div(category_region.sum(axis=1), axis=0) * 100
    category_region_pct.plot(kind='bar', stacked=True, ax=ax4, 
                              color=[COLORS['primary'], COLORS['secondary'], COLORS['accent']], 
                              alpha=0.8)
    ax4.set_xlabel('Region', fontweight='bold')
    ax4.set_ylabel('Percentage (%)', fontweight='bold')
    ax4.set_title('Product Category Mix by Region', pad=15)
    ax4.legend(title='Category', bbox_to_anchor=(1.05, 1), loc='upper left')
    ax4.grid(True, alpha=0.3, axis='y')
    ax4.tick_params(axis='x', rotation=0)
    
    return fig

def save_all_visualizations():
    """Generate and save all visualizations"""
    os.makedirs(OUTPUT_PATH, exist_ok=True)
    
    # Load data
    df = load_data()
    if df is None:
        return
    
    print("\nGenerating Python visualizations...")
    print("="*60)
    
    # Generate and save plots
    plots = {
        'revenue_heatmap_py.png': plot_revenue_heatmap,
        'sales_distribution_py.png': plot_sales_distribution,
        'time_series_advanced_py.png': plot_time_series_advanced,
        'regional_comparison_py.png': plot_regional_comparison,
        'category_performance_py.png': plot_category_performance
    }
    
    for filename, plot_func in plots.items():
        try:
            fig = plot_func(df)
            filepath = os.path.join(OUTPUT_PATH, filename)
            fig.savefig(filepath, dpi=300, bbox_inches='tight', facecolor='white')
            plt.close(fig)
            print(f"✓ Saved: {filename}")
        except Exception as e:
            print(f"✗ Error creating {filename}: {str(e)}")
    
    print(f"\nAll visualizations saved to {OUTPUT_PATH}/")
    print("="*60)

def main():
    """Main execution"""
    save_all_visualizations()

if __name__ == "__main__":
    main()
